stdlib/public/core/IntegerParsing.swift

```diff
@@ -10,15 +10,404 @@
+  ///     value. `radix` must be in the range `2...36`. The default is 10.
+  @inlinable @inline(__always)
+  public init?<S : StringProtocol>(_ text: S, radix: Int = 10) {
+    precondition(2...36 ~= radix)
+    let result_: Self?
+    if let text = text as? String, text._guts.isSmall {
+      result_ = _parseSmall(text._guts.rawBits, radix: radix)
+    } else {
+      result_ = text.utf8.withContiguousStorageIfAvailable { utf8 -> Self? in
+        var iter = utf8.makeIterator()
+        return _parseFromUTF8Inlined(&iter, radix: radix)
+      } ?? {
+        var iter = text.utf8.makeIterator()
+        return _parseFromUTF8(&iter, radix: radix)
+      }()
+    }
+    guard _fastPath(result_ != nil), let result = result_ else { return nil }
+    self = result
+  }
+
+  /// Creates a new integer value from the given string.
+  ///
+  /// The string passed as `description` may begin with a plus or minus sign
+  /// character (`+` or `-`), followed by one or more numeric digits (`0-9`).
+  ///
+  ///     let x = Int("123")
+  ///     // x == 123
+  ///
+  /// If `description` is in an invalid format, or if the value it denotes in
+  /// base 10 is not representable, the result is `nil`. For example, the
+  /// following conversions result in `nil`:
+  ///
+  ///     Int(" 100")                       // Includes whitespace
+  ///     Int("21-50")                      // Invalid format
+  ///     Int("ff6600")                     // Characters out of bounds
+  ///     Int("10000000000000000000000000") // Out of range
+  ///
+  /// - Parameter description: The ASCII representation of a number.
+  @inlinable
+  public init?(_ description: String) {
+    self.init(description, radix: 10)
+  }
+}
+
+@inlinable @inline(__always)
+internal func _parseSmall<Result : FixedWidthInteger>(
+  _ rawGuts: _StringObject.RawBitPattern, radix: Int
+) -> Result? {
+  // We work with the bytes as they are arranged in the StringObject,
+  // i.e. the leading characters are in the least-significant bits.
+  // This is as opposed to SmallString order, which swaps them on big-endian
+  // platforms to maintain the same in-memory order.
+  var word1 = rawGuts.0
+  let word2 = rawGuts.1 & 0x00ff_ffff_ffff_ffff
+  let count = Int((rawGuts.1 &>> 56) & 0x0f)
+  // Handle sign.
+  // "0"..."9" == 0x30...0x39
+  // "-" == 0x2d
+  // "+" == 0x2b
+  var hasMinus = false
+  let first = word1 & 0xff
+  if first < 0x30 { // Plus, minus or an invalid character.
+    guard _fastPath(count > 1), // Disallow "-"/"+" without any digits.
+      _fastPath(Result.isSigned && first == _ascii("-"))
+      || _fastPath(first == _ascii("+"))
+      else { return nil }
+    hasMinus = _fastPath(first == _ascii("-"))
+    word1 = (word1 ^ first) | 0x30 // Clear first char and a "0".
+  } else {
+    guard _fastPath(count > 0) else { return nil }
+  }
+  // Choose specialization based on radix (normally branch is optimized away).
+  let result_: UInt64?
```

Comment:
Given this, the implementation can't actually be generic over `Result: FixedWidthInteger`, because it doesn't work for a conforming type larger than `UInt64`, right?

Reply:
It's fine for base 2/10/16, since the max number of characters in a small string is 15 and 16**15 < 2**64. The default case however could be as large as 36**15 so that one indeed shouldn't use that intermediate.